### PR TITLE
Po spage

### PR DESCRIPTION
**Pull request by RemaaBdair**

### CODE REVIEW

File: src/reducers/index.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+const reducer = createReducer<State>({}, defaultState);
+
+reducer.on(addItem, (state, payload) => ({
+  ...state,
+
+  items: [...state.items, payload],
+}));
+reducer.on(fetchProductsResult, (state: State, payload: Product[]) => {
```
Comment (Abdallatif) [inline src/reducers/index.tsx:28]: let's separate the reducers each on it's own and then we can combine them using `combineReducers` from redux. The reason for this is separation, we want to keep different reducers/logic separated

File: src/store/saga.tsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { call, put, takeEvery } from "redux-saga/effects";
+import { Product } from "../pages/Products/util";
+import { Category } from "../pages/Category/util";
+export const api = (url: string) =>
+  fetch(url).then((response) => response.json());
+export function* fetchProducts() {
```
Comment (Abdallatif) [inline src/store/saga.tsx:6]: same as reducers, separate the sagas

File: src/store/saga.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    yield put({
+      type: "FETCH_ALL_CATEGORIES_ON_SUCCESS",
+      payload: categories,
+    });
+  } catch (e) {
+    console.log(e);
+  }
+}
+export function* rootSaga() {
+  yield takeEvery("FETCH_ALL_PRODUCTS_REQUEST", fetchProducts);
```
Comment (Abdallatif) [inline src/store/saga.tsx:35]: you may want to use `takeLatest` here

File: src/pages/POS/Components/StockItems/ItemsList.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            </div>
+          </Card>
+        );
+      })}
+    </Grid>
+  );
+};
+
+const mapStateToProps = (state: State) => ({
+  products: state.products,
```
Comment (Abdallatif) [inline src/pages/POS/Components/StockItems/ItemsList.tsx:66]: it's a good pattern to have a file called `selectors` where each selector take the state as input and return the value from the state, like:
```
// selectors.ts
const selectProducts = (state: State) => state.products

// ItemsList.tsx
const mapStateToProps = (state: State) => ({
  products: selectProducts(state),
})
```
By this, the component doesn't need to know what's the shape of the state and if you change the structure of the state, you only update the selector in one place

File: src/pages/POS/Components/StockItems/ItemsList.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            key: "category",
+          },
+        ],
+        result
+      ),
+    [result, categoryFilter]
+  );
+  const handleClick = (product: Product) => {
+    addItem(product);
+    console.log("hi");
```
Comment (Abdallatif) [inline src/pages/POS/Components/StockItems/ItemsList.tsx:38]: remove console.log

File: src/pages/POS/Components/StockItems/CategoryFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+interface Props {
+  activeFilter: string | null; //category perhaps
+  categories: Category[];
+  onClick: (categoryName: string | null) => void;
+}
+export const CategoryFilter: React.FunctionComponent<
+  WithStyles<typeof styles> & Props
+> = (props) => {
+  const { classes, activeFilter, categories, onClick } = props;
+  console.log(categories);
```
Comment (Abdallatif) [inline src/pages/POS/Components/StockItems/CategoryFilter.tsx:17]: remove console.log

File: src/pages/POS/Components/StockItems/StockItems.tsx [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        />
+      </Grid>
+      <Grid item xs={12} justify="flex-start" container>
+        <ItemsList searchText={searchText} categoryFilter={categoryFilter} />
+      </Grid>
+    </Grid>
+  );
+};
+const mapStateToProps = (state: State) => ({
+  categories: state.categories,
```
Comment (Abdallatif) [inline src/pages/POS/Components/StockItems/StockItems.tsx:65]: use selectors

File: src/store/saga.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import { Category } from "../pages/Category/util";
+export const api = (url: string) =>
+  fetch(url).then((response) => response.json());
+export function* fetchProducts() {
+  try {
+    const products: Product[] = yield call(
+      api,
+      "http://localhost:3001/products"
+    );
+    yield put({
```
Comment (Abdallatif) [inline src/store/saga.tsx:12]: you can `yield put(fetchProductsResult(products))`

File: src/store/saga.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
+import { call, put, takeEvery } from "redux-saga/effects";
+import { Product } from "../pages/Products/util";
+import { Category } from "../pages/Category/util";
+export const api = (url: string) =>
+  fetch(url).then((response) => response.json());
+export function* fetchProducts() {
+  try {
+    const products: Product[] = yield call(
+      api,
```
Comment (Abdallatif) [inline src/store/saga.tsx:9]: it's better to have this as separate function `yield call(fetchProducts)`

File: src/store/saga.tsx (hunk 3 above)
Comment (RemaaBdair) [reply src/store/saga.tsx:35]: but it won't make a difference here as we are dispatching this action once, right?

Comment (Abdallatif) [reply src/store/saga.tsx]: yes, but the idea we don't know who/when the action is dispatched

File: src/pages/POS/Components/StockItems/CategoryFilter.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  const { classes, activeFilter, categories, onClick } = props;
+  return (
+    <Grid container alignItems="flex-start">
+      <Button
+        id="home"
+        type="button"
+        variant="outlined"
+        onClick={() => onClick(null)}
+        classes={
+          activeFilter === null
```
Comment (Abdallatif) [inline src/pages/POS/Components/StockItems/CategoryFilter.tsx:25]:
```suggestion
         { root: activeFilter === null ? classes.activeFilter : classes.inActiveFilter}
```

File: src/store/index.tsx [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+const sagaMiddleware = createSagaMiddleware();
+const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
+
+const store = createStore(
+  rootReducer,
+  composeEnhancers(applyMiddleware(sagaMiddleware))
+);
+
+sagaMiddleware.run(rootSaga);
+store.dispatch({ type: "FETCH_ALL_PRODUCTS_REQUEST" });
```
Comment (Abdallatif) [inline src/store/index.tsx:32]: dispatch these actions on pos page mount, so we don't fetch it in other pages where we don't want it

File: src/actions/index.tsx [hunk 12]
```diff
@@ -0,0 +1,11 @@
+import { createAction } from "redux-act";
+import { Product } from "../pages/Products/util";
+import { Category } from "../pages/Category/util";
+
+export const addItem = createAction<Product>("add item to the cart");
+export const fetchProductsResult = createAction<Product[]>(
+  "FETCH_ALL_PRODUCTS_ON_SUCCESS"
```
Comment (Abdallatif) [inline src/actions/index.tsx:7]: "fetch products success", it's easier to read
```suggestion
  "fetch products success"
```

File: src/store/index.tsx (hunk 11 above)
Comment (Abdallatif) [reply src/store/index.tsx:32]: and make this an action

File: src/reducers/cart.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+const defualtState = {
+  products: [],
+};
+const reducer = createReducer<Cart>({}, defualtState as any);
+reducer.on(addProduct, (state, payload: Product) => {
+  let isFound: boolean = false;
+  const filteredProducts = state.products.filter((elem) => {
+    if (elem.id === payload.id) {
+      elem.quantity += 1;
```
Comment (Abdallatif) [inline src/reducers/cart.tsx:20]: don't mutate the state

File: src/reducers/cart.tsx [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      ...state,
+
+      products: filteredProducts,
+    };
+});
+
+reducer.on(incQuantity, (state, payload: string) => {
+  const filteredProducts = state.products.filter((elem) => {
+    if (elem.id === payload) {
+      elem.quantity += 1;
```
Comment (Abdallatif) [inline src/reducers/cart.tsx:42]: state mutation

File: src/reducers/cart.tsx [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    ...state,
+
+    products: filteredProducts,
+  };
+});
+
+reducer.on(decQuantity, (state, payload: string) => {
+  const filteredProducts = state.products.filter((elem) => {
+    if (elem.id === payload && elem.quantity > 1) {
+      elem.quantity -= 1;
```
Comment (Abdallatif) [inline src/reducers/cart.tsx:57]: state mutation

File: src/reducers/items.tsx [hunk 16]
```diff
@@ -1,12 +1,35 @@
...
+const reducer = createReducer<state>({}, {} as any);
+reducer.on(addProduct, (state, payload: Product) => {
+  let isFound: boolean = false;
+  for (let x in state) {
+    if (x === payload.id) {
+      isFound = true;
+    }
+  }
+  if (!isFound) {
+    state[payload.id] = {
```
Comment (Abdallatif) [inline src/reducers/items.tsx:17]: state mutation

File: src/reducers/items.tsx [hunk 17]
```diff
@@ -1,12 +1,35 @@
...
+      price: +payload.price.slice(0, -1),
+    };
+  }
 
+  return state;
+});
+reducer.on(deleteProduct, (state, payload: string) => {
+  const products = state;
+  for (let x in products) {
+    if (x === payload) delete products[x];
```
Comment (Abdallatif) [inline src/reducers/items.tsx:28]: state mutation

File: src/reducers/items.tsx [hunk 18]
```diff
@@ -1,12 +1,35 @@
...
+  }
+  return products;
+});
+reducer.on(deleteCart, (state) => {
+  return [] as any;
```
Comment (Abdallatif) [inline src/reducers/items.tsx:33]: no any

**Pull request merged**